cmd/controller/run.go

```diff
@@ -63,13 +64,26 @@ func run(options *config.ControllerOptions) error {
 	}
 	s := server.NewHTTPServer(&httpServerOptions)
 	s.RegisterControllerHandlers()
-	go s.Run()
 
-	if noLeaderElection {
-		callback(ctx)
-	} else {
-		leader.RunOrDie(ctx, "kube-system", "vm-dhcp-controllers", client, callback)
-	}
+	eg, egctx := errgroup.WithContext(ctx)
+
+	eg.Go(func() error {
+		return s.Run()
+	})
+
+	eg.Go(func() error {
+		<-egctx.Done()
+		return s.Stop(egctx)
+	})
+
+	eg.Go(func() error {
+		if noLeaderElection {
+			callback(egctx)
+		} else {
+			leader.RunOrDie(egctx, "kube-system", "vm-dhcp-controllers", client, callback)
+		}
+		return nil
+	})
 
-	return nil
+	return eg.Wait()
```

Comment:
I'm thinking should we wait `<-egctx.Done()` inside `eg.Go`?

For example, `eg.Wait` waits forever because no one calls cancel, and egctx will not become canceled because `<-egctx.Done()` is pending. It's like circular.

```go
func main() {
	ctx, _ := context.WithCancel(context.Background())

	group, egctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		return nil
	})

	group.Go(func() error {
		<-egctx.Done()
		fmt.Println("do something")
		return nil
	})

	if err := group.Wait(); err != nil {
		log.Println("error is ", err)
	}

	log.Println("done")
}
```

Reply:
I think the `cancel()` is called [here](https://github.com/rancher/wrangler/blob/004384a454ec6062420e10439bf9fa136fa38e4d/pkg/signals/signal.go#L46) once a signal is received.